### PR TITLE
DOC: Use napoleon sphinx extension

### LINKED ISSUE #5731: Move from numpydoc to sphinxcontrib-napoleon
napoleon is a more general replacement for numpydoc docstring support and ships with Sphinx since 1.3.

There seem to be subtle differences, however, and so it will take some manual updating to our docstrings to get this to work.

### PR DESCRIPTION
**Pull request by QuLogic**
This is now building without warning, but there are a few differences. For some reason, some images are not consistent between this and `v2.x`; I think some of the documentation scripts are missing sorting (I'm using Python 3).

The biggest difference seems to be with parameter listings. The existing result seems to be a bunch of paragraphs and blockquotes, but the new result is a bunch of bullet lists. I think this may actually be more correct, but it does change the look somewhat.

Fixes #5731.

### CODE REVIEW

**Comment by QuLogic** [comment]
Also, I'm wondering what's the point of some of the "Call signature" lines in `axes/_axes.py` because they seem to be redundant?

**Comment by tacaswell** [comment]
The `Call signature` lines probably pre-date sphinx.

**Comment by efiring** [comment]
It looks like most of those `Call signature` blocks can be deleted, but there are exceptions such as `barbs`, `quiver`, `contour`, and `clabel` where there are multiple signatures hidden within `*args` that Sphinx cannot handle automatically.

**Comment by QuLogic** [comment]
I spent a little time looking into the different images, and it seems like, e.g., the tight layout demo, just picks font sizes at random. It makes the doc build a bit unreproducible; can we stop doing that?

**Comment by mdboom** [comment]
> I spent a little time looking into the different images, and it seems like, e.g., the tight layout demo, just picks font sizes at random. It makes the doc build a bit unreproducible; can we stop doing that?

Yes!  In this case, I don't see any reason why it's random, let's just hardcode the values in the example.  In other cases where the example really does want to use random data, we should at least seed the random number generator so the results are predictable.

**Comment by tacaswell** [comment]
This needs a rebase.

**Comment by QuLogic** [comment]
Yes, waiting for #5769 as well.

File: doc/README.txt [hunk 1]
```diff
@@ -33,11 +33,11 @@ python documentation system built on top of ReST.  This directory contains
...
-To build the HTML documentation, install sphinx (1.0 or greater
-required), then type "python make.py html" in this directory.  Wait
-for the initial run (which builds the example gallery) to be done,
-then run "python make.py html" again. The top file of the results will
-be ./build/html/index.html
+To build the HTML documentation, install sphinx (1.0 or greater required) and,
+if sphinx older than 1.3, sphinxcontrib-napoleon, then type "python make.py
+html" in this directory.  Wait for the initial run (which builds the example
+gallery) to be done, then run "python make.py html" again. The top file of the
+results will be ./build/html/index.html
```
Comment (jenshnielsen) [inline doc/README.txt]: Do we really still need to run it twice? I think this is left over from some old setup which should be removed. Travis only runs python make once and that seems to work fine.

Comment (QuLogic) [inline doc/README.txt]: Yea, I wondered about that, but I don't know for sure why it's there.

Comment (mdboom) [inline doc/README.txt]: Yeah -- I had forgotten that was there.  I vaguely recall that was once true (back in ancient times when we didn't even use Sphinx extensions), but I haven't had to do that in years.  Let's take it out while we're in here.

Comment (QuLogic) [inline doc/README.txt]: Alright, I removed it.

**Comment by QuLogic** [comment]
I was going to add in fixes to several docstrings with incorrect headings and such, but that might balloon this PR to something rather big. The changes here are enough to get the build to pass without error, so maybe that should be left to some later PR?

**Comment by tacaswell** [comment]
Better to not let the prefect be the enemy of the good.

**Comment by QuLogic** [comment]
Let's call this done for now then, and I'll work on the other stuff in another PR as time permits.

**Comment by mdboom** [comment]
Should we hold off on this until investigating the bugs @QuLogic alluded to on gitter?

**Comment by QuLogic** [comment]
Once we merge the PRs from @jenshnielsen, this should work without warning on Sphinx 1.4. The `napoleon_use_param=False` setting also makes the bugs go away with 1.3, but there are still some warnings.

I don't know whether or not you still wish to put this on 2.0, but it shouldn't break anything at that point.

**Comment by QuLogic** [comment]
With the recent PRs merged, this one should work correctly with Sphinx 1.4.0, but for some reason Travis doesn't want to install it. Anyway, because #6243 sets warnings to be errors by default, this build fails with Sphinx 1.3.1.

**Comment by jenshnielsen** [comment]
On 1.5.x and 2.x Travis installs most dependencies from the scikit image wheel house which only has specific prebuild versions of the dependencies. On master only Pandas and Numpy (which have significant amount of compiled code) are installed from the wheel house. That change was never backported but there is no reason that Sphinx has to be installed from the wheel house on 2.x

#6243 does not set warnings as errors per default on Travis. That has been on for a long time. It only inverst the logic such that you have to pass a flag to disable warnings as errors rather than passing a flag to enable it. Before that pr travis passed --warningsaserrors to make.py

**Comment by QuLogic** [comment]
> #6243 does not set warnings as errors per default on Travis. That has been on for a long time.

Yes, sorry, you're right. I had forgotten that it did so.

**Comment by NelleV** [comment]
I am not sure why I was assigned to this PR, so I have unassigned myself.
I actually think we should stick to numpydoc, considering it is the standard in the scientific python community. It is much easier for us to push changes to numpydoc than it is to napoleon.

**Comment by jklymak** [comment]
Is this something that will ever happen?  Folks have been putting a lot of documentation effort in recently.  If the standard is supposed to be slightly different it'd be good to know sooner rather than later.  If this won't happen, this can be closed?

**Comment by efiring** [comment]
There are two parts to this: docstring cleanups and the switch to Napoleon.  I suspect most of the docstring cleanups have been done by now in other PRs, but I haven't checked.  As for the switch to Napoleon, the advantage is that it removes a dependency; the disadvantage is that Napoleon might not perfectly track numpydoc.
@QuLogic, what are your thoughts on this now?

**Comment by timhoffm** [comment]
Indeed, most docstring cleanups from here are already merged. #11064 handles the remaining docstring changes, so that this PR can be significantly simplified and focus only on a possible switch to napoleon.

**Comment by jklymak** [comment]
Closing as it doesn't seem there is any appetite to switch to napoleon.  OTOH, if someone wants to take the "switch" part of the PR on, that'd be fine w/ me....

**Comment by anntzer** [comment]
Reopening per https://github.com/matplotlib/matplotlib/pull/11397#discussion_r195863999.

**Comment by timhoffm** [comment]
If we want to switch to napoleon, we'd need a careful look at the changes that may introduce in the generated files. While technically implementing the same specification, napoleon and numpydoc interpret it differently, sometimes stricter, sometimes more strict. We're probably knowingly and unknowingly adapting our docstrings to the numpydoc interpretation by just using it and checking what works and what doesn't. This "learned" formatting has to be reevaluated when swiching to napoleon.

**Comment by anntzer** [comment]
fwiw, it looks like sklearn (including one of numpydoc's maintainers...) is considering switching to napoleon (https://github.com/scikit-learn/scikit-learn/issues/11440).

**Comment by matthew-brett** [comment]
But - reading that thread - they are pretty ambiguous about it.  The main argument for Napoleon seems to be 'it comes with sphinx' - but that's not a very powerful argument on its own, given that numpydoc is so easy to install.

**Comment by anntzer** [comment]
That's why I said "considering", not "decided".
However I agree with Gael's points:

> Is there a long-term reason that Napoleon would get less maintenance than
numpydocs if projects switch to it? It's embedded in sphinx, so it will
get a lot of exposure. Of course, it's a question of community dynamics,
and there is a lot of social in there.

> That it's included in sphinx. I tend to hope that it is more likely to
become a community standard. Everybody who does sphinx docs has napoleon.

**Comment by matthew-brett** [comment]
Sure, but I don't know if it's really true the Napoleon really gets better maintenance.   Among the big scientific Python projects, Numpy, Scipy, Pandas, Scikit-image, Scikit-learn, Sympy and Astropy all use Numpydoc, with Statsmodels being the only one I know of that uses Napoleon.  Pauli Virtanen leads Scipy, and maintains Numpydoc, so I think it's reasonable to hope it will stay fit for purpose for a while.   How about waiting with a plan to review in a year's time?

**Comment by timhoffm** [comment]
There are some subtle differences in the parsers concerning name and type detection and formatting.

e.g. https://github.com/numpy/numpydoc/issues/72#issuecomment-386143335

We now tune the docs to be numpydoc compliant. So we really would have to test how napoleon behaves. Nevertheless, I have the feeling that napoleon has a more sane parser.

**Comment by dstansby** [comment]
I *supsect* even if we want to switch to `napoleon`, this PR is so out of date that attempting to rebase it will be more hassle than it's worth; if no-one objects in a week or so I'll close this but leave the issue open.
